README.md
# Performance Signature with Dynatrace (AppMon) for Jenkins
[![Build Status](https://ci.jenkins.io/job/Plugins/job/performance-signature-dynatrace-plugin/job/master/badge/icon)](https://ci.jenkins.io/job/Plugins/job/performance-signature-dynatrace-plugin/job/master/)
[![Jenkins Plugin](https://img.shields.io/jenkins/plugin/v/performance-signature-dynatrace.svg)](https://plugins.jenkins.io/performance-signature-dynatrace/)
[![Changelog](https://img.shields.io/github/v/tag/jenkinsci/performance-signature-dynatrace-plugin?label=changelog)](https://github.com/jenkinsci/performance-signature-dynatrace-plugin/blob/master/CHANGELOG.md)
[![Jenkins Plugin Installs](https://img.shields.io/jenkins/plugin/i/performance-signature-ui?color=blue)](https://plugins.jenkins.io/performance-signature-dynatrace/)

Smoking latest software products are created with the help of flexible and agile systems in a continuous integration (CI) environment. Such environments include
a CI-Server like Jenkins. The Performance Signature collects performance values during a build and evaluates and compares the result with previous builds and
non-functional requirements. Several software tests can be run and evaluated automatically, so that the most important key performance indicators (KPI) can be
summarized and be available for all project participants very quickly.

### Get in touch with Application Performance Experts

![T-Systems](https://www.t-systems-mms.com/typo3conf/ext/mkmms/Resources/Public/img/logos/logo.png)

**T-Systems Multimedia Solutions GmbH**

Send a [mail](mailto:performance@t-systems-mms.com) to the Application Performance Team

### Awards
* Winner of the Dynatrace Solution Innovation Award EMEA 2016
* Winner of the Dynatrace EMEA API Challenge 2018

---
### Table of Contents

<!-- toc -->

- [Introduction](#introduction)
- [Installation](#installation)
  * [Using Jenkins Update Center](#using-jenkins-update-center)
  * [Manual Installation](#manual-installation)
- [Problems? Questions? Suggestions?](#problems-questions-suggestions)
- [Additional Resources](#additional-resources)
  * [Dynatrace AppMon Documentation](#dynatrace-appmon-documentation)
  * [Dynatrace SaaS/Managed Documentation](#dynatrace-saasmanaged-documentation)
  * [Links](#links)

<!-- tocstop -->

## Introduction

The Performance Signature evaluates aggregated load and performance data from Dynatrace after each build.
We are currently supporting three data sources:
* Dynatrace Application Monitoring (AppMon)
* Dynatrace SaaS/Manged
* Remote Jenkins Jobs with enabled Performance Signature

Each data source has it's own Jenkins plugin, whereof each one depends on the Performance Signature UI Plugin.
Find below the documentation of each component.

* **[Performance Signature: Dynatrace AppMon](dynatrace-appmon/README.md)**
* **[Performance Signature: Dynatrace SaaS/Managed](dynatrace/README.md)**
* **[Performance Signature: Dynatrace Viewer](viewer/README.md)**

## Installation
### Using Jenkins Update Center

The recommended way of installing the plugin is by using the Update Center (plugin directory). Navigate to `Manage Jenkins -> Manage Plugins` page and switch to the `Available` tab. Search for the keyword "Performance Signature:" and install the plugin.

### Manual Installation

This procedure is meant for developers, who want to install a locally built plugin version.

* build the plugin from source using `mvn package` command
* in Jenkins, go to the `Manage Jenkins -> Manage Plugins`
* switch to the `Advanced` tab
* upload the built plugin package from the `target/performance-signature-*.hpi` path to the `Upload Plugin` section
* restart Jenkins

## Problems? Questions? Suggestions?

Contact the Performance Experts at T-Systems: `performance@t-systems-mms.com` for professional support or consultancy. T-Systems offers support packages starting at 500 € p.a.

## Additional Resources

### Dynatrace AppMon Documentation

- [Continuous Delivery & Test Automation](https://community.dynatrace.com/community/pages/viewpage.action?pageId=215161284)
- [Capture Performance Data from Tests](https://community.dynatrace.com/community/display/DOCDT63/Capture+Performance+Data+from+Tests)
- [Integrate Dynatrace in Continuous Integration Builds](https://community.dynatrace.com/community/display/DOCDT63/Integrate+Dynatrace+in+Continuous+Integration+Builds)

### Dynatrace SaaS/Managed Documentation

- https://www.dynatrace.com/support/help/

### Links

* [Dynatrace Plugin Page](https://community.dynatrace.com/community/display/DL/Performance+Signature+Plugin)
* [Application Performance Team @ T-Systems](https://test-and-integration.t-systems-mms.com/leistungen/application-performance-management.html)
* [Open issues](https://issues.jenkins-ci.org/issues/?jql=project%20%3D%20JENKINS%20AND%20status%20in%20(Open%2C%20%22In%20Progress%22%2C%20Reopened)%20AND%20component%20%3D%20%27performance-signature-dynatrace-plugin%27)
* [Changelog](CHANGELOG.md)


CHANGELOG.md
Changelog
====

### Version 3.2.2 (Jun 07, 2021)

* use okhttp dependency from io.jenkins.plugin:okhttp-api plugin
* optimize ui
* use font-awesome-api plugin and remove old css

### Version 3.2.1 (Apr 15, 2021)

* fixed [JENKINS-65184 & JENKINS-56109] Plugins breaking "Configure System" UI
* fixed [JENKINS-64396] remove Dynatrace server button label mislabeled
* fix NPEs in Dynatrace module
* rebuild UI with webpack and gridstack
* several dependency updates
* increase minimum jenkins version to 2.277.1
* Dev: update parent pom version to 4.18
* Dev: bump bom-2.264.x to version 28

### Version 3.2.0 (Apr 15, 2021)

skipped due to issues with the release process

### Version 3.1.6 (Nov 14, 2020)

* fix NPE in AppMon module
* update url to diagnostic tools
* Optimized Images [(9b621c3)](https://github.com/jenkinsci/performance-signature-dynatrace-plugin/commit/d268da229cd2c15eb5515e27ea23bdadae09448c)
* several dependency updates
* increase minimum jenkins version to 2.164.3
* Dev: update parent pom version to 4.13
* Dev: bump bom-2.164.x from 9 to 10

### Version 3.1.5 (Jan 16, 2020)
* JENKINS-55202 fix classloading issues with JAXB
* log error message, instead of response message
* increase minimum jenkins version to 2.164.1
* several dependency updates
* Dev: update parent pom version to 3.55
* Dev: switch to wiremock for Dynatrace AppMon

### Version 3.1.4 (Jun 18, 2019)
* activated [Dependabot](https://dependabot.com/)
* several dependency updates
* Dev: update parent pom version to 3.46

### Version 3.1.3 (Apr 16, 2019)
* added a DynatraceCustomSession pipeline step for fixed timeframes (can be found under the advanced section)
* fixed NPE in DynatraceReportStepExecution
* Dev: update parent pom version to 3.42

### Version 3.1.2 (Mar 26, 2019)
* use build number instead of display name
* use aggregation from spec definition instead of fixed avg aggregation
* filter out "Synthetic Web Requests by Timer Name - PurePath Response Time - " from measure name
* dev: updated java dependencies

### Version 3.1.1 (Mar 11, 2019)
* increased connect & read timeout
* don't check for session recording status for production licenced instances
* added a dependency to jaxb-plugin for java 11 compatibility
* updated parent pom, nodejs & yarn version to latest version
* Jenkins minimum version is now 2.89.4

### Version 3.1.0 (Jan 24, 2019)
* fix unit conversion (fixed time unit: MilliSecond, fixed byte unit: MegaByte)
* introduced new SpecFile format
  * global lowerLimit & upperLimit values are not supported anymore
  * lowerLimit & upperLimit are deprecated
  * introduced lowerWarning, lowerSevere, upperWarning, upperSevere
* print out more readable error messages

### Version 3.0.3 (Jan 18, 2019)
* [JENKINS-55562] - Apply workaround in the version number
* always use Instant.now().toEpochMilli() to get a UTC timestamp
* create a separate specfile for draft version 2.0
* added custom properties to `CreateDynatraceDeployment` step

### Version 3.0.2 (Jan 03, 2019)
* fix CVE-2018-1000850 and CVE-2018-1000844
* drop analysis-core dependency
* fix NPE in round method
* whitelist some methods in PerfSigEnvInvisAction class

### Version 3.0.1 (Sep 14, 2018)
* fixed release problems

### Version 3.0.0 (Sep 13, 2018)
* new module: Performance Signature for Dynatrace Saas/Managed
* Requires Jenkins Core 2.60.3 or newer
* Requires Java 8
* complete overhaul of the http client
* removed custom proxy option
* security hardening of all web methods
* group up all incidents by severity and message
* be more verbose about network related exception
* changed argument values for parameter 'nonFunctionalFailure'

### Version 2.5.8 (Jul 04, 2018)
* complete rewrite of the Performance Signature viewer module
* Dev: use yarn instead of npm
* Dev: update parent pom version to 3.17

### Version 2.5.7 (May 17, 2018)
* major Bugfix: don't use a static instance of ApiClient
* Dev: code cleanup
* Dev: update parent pom version to 3.8

### Version 2.5.6 (Mar, 26, 2018)
* Bugfix: Grid arrangement gets mixed up
* Bugfix: delete only the previous session
* mark pipeline step dependency as optional
* Dev: update parent pom version to 3.6

### Version 2.5.2 (Jan 17, 2018)
* Bugfix: Viewer configuration was not saved
* Bugfix: removed minimum dashboard requirement from jelly
* Feature: added option to delete sessions after reporting and session export
* Dev: update parent pom version to 3.2
* Dev: update gridster JS lib to latest version

### Version 2.5.1 (Dec 22, 2017)
* Bugfix: empty chart configuration caused by missing alias

### Version 2.5.0 (Dec 21, 2017)
* only compatible with Dynatrace AppMon 7.0 and higher
* added "Create Deployment" Pipeline Step
* all log output is now prefixed by "[PERFSIG]"
* complete rewrite of rest interface by using a generated java client 
* reordered Session Recording and TestRun creation/finish
* reanalyse Session Recording option is not available anymore
* simplified test data persistence
* actual test run id can be accessed via "DYNATRACE_TESTRUN_ID"
* confidential strings can now be removed (default: true)
* read timeout can be configured (default: 300s)
* aligned wording of plugin names
* use JAXB for xml parsing
* show incidents within the Performance Signature Build report
* huge internal code clean up

### Version 2.4.4 (Sep 13, 2017)
* replaced protocol, host and port with REST endpoint URL in global configuration
* fix a visual bug by using Pipeline: Stage View Plugin and Performance Signature Plugin in a pipeline job
* internal code clean up

### Version 2.4.3 (Jun 07, 2017)
* Java 7 is now a minimum requirement
* updated JS dependencies
* sort all ListBoxes
* fix PerfSigViewer remote build step

### Version 2.4.2 (Mar 28, 2017)
* updated dependencies
* avoid duplicated charts by using dynamic measures (e.g. application splitting)
* added a random parameter, so charts don't get cached

### Version 2.4.1 (Jan 26, 2017)
* refactored log messages
* added aggregation to chart title

ui/src/main/js/projectaction.js
/*
 * Copyright (c) 2014-2018 T-Systems Multimedia Solutions GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import wurl from 'wurl';
import 'lightbox2';
import {GridStack} from 'gridstack';
import 'gridstack/dist/gridstack.css';
import 'gridstack/dist/h5/gridstack-dd-native';
import 'lightbox2/dist/css/lightbox.css';

(function ($) {
    'use strict';
    let grid = GridStack.initAll({
        margin: 5,
        cellHeight: 140,
    }, '.grid-stack');

    let randomParam = '&_=' + $.now();
    $.fn.dataTableExt.sErrMode = 'none';

    $('.panel-body').each(function () {
        if (!/[\S]/.test($(this).html())) {
            $(this).html('no PDF reports available!');
        }
    });

    $('.tab-pane').each(function (pageIndex, page) {
        try {
            $('.table', this).has('tbody').dataTable({
                'stateSave': false,
                'order': [0, 'desc']
            });
        } catch (e) {
            console.log(e);
        }

        $('#measureGroup', this).change(function () {
            if ($(this).val() === 'UnitTest overview') {
                $('#measure', page).parent().hide();
                $('#aggregation', page).parent().hide();
                $('#customName', page).val('');
            } else {
                projectAction.getAvailableMeasures($(page).attr('id'), $(this).val(), function (data) {
                    $('#measure', page).empty();
                    $('#customName', page).val('');
                    $('#customBuildCount', page).val('');
                    $('#measure', page).parent().show();
                    $('#aggregation', page).parent().show();
                    $.each(data.responseObject(), function (val, text) {
                        $('#measure', page).append($('<option></option>').val(val).html(text));
                    });
                    $('#measure', page).trigger('change');
                });
            }
        });

        $('#measure', this).change(function () {
            projectAction.getAggregationFromMeasure($(page).attr('id'), $('#measureGroup', page).children(':selected').text(),
                $(this).children(':selected').text(), function (data) {
                    $('#aggregation', page).val(data.responseObject());
                    $('#aggregation', page).trigger('change');
                });
        });

        $('#aggregation', this).change(function () {
            $('#customName', page).val(generateTitle($('#measure', page).children(':selected').text(), $('#measureGroup', page).children(':selected').text(),
                $(this).children(':selected').text()));
        });

        $('#editbutton', this).click(function () {
            $(this).hide();
            $('.img-thumbnail', page).unwrap();
            $('#measureGroup', page).trigger('change');
            $('#donebutton', page).show();
            $('#cancelbutton', page).show();
            $('#editform', page).show();
            $('.del_img', page).show();
            $('.chk_show', page).show();
            grid[pageIndex].enableMove(true);
        });
        $('#cancelbutton', page).click(function () {
            location.reload(true);
        });

        $('#addbutton', page).click(function () {
            const request_parameter = '&amp;width=410&amp;height=300&amp;customName=' + encode($('#customName', page).val()) +
                '&amp;customBuildCount=' + $('#customBuildCount', page).val();
            if ($('#measureGroup', page).val() === 'UnitTest overview') {
                grid[pageIndex].addWidget({
                    w: 3, h: 2, content: `<span class="del_img float-left">
<svg class="svg" style="fill: #ff0000" aria-hidden="true">
    <use href="${resURL}/plugin/font-awesome-api/sprites/solid.svg#times"/>
</svg></span>
<span class="chk_show float-right">
<input type="checkbox" title="show in project overview" checked="checked"/></span>
<img class="img-thumbnail" height="300" width="410" src="testRunGraph?id=unittest_overview${request_parameter}${randomParam}">`
                });
            } else {
                grid[pageIndex].addWidget({
                    w: 3, h: 2, content: `<span class="del_img float-left">
<svg class="svg" style="fill: #ff0000" aria-hidden="true">
    <use href="${resURL}/plugin/font-awesome-api/sprites/solid.svg#times"/>
</svg></span>
<span class="chk_show float-right">
<input type="checkbox" title="show in project overview" checked="checked"/></span>
<img class="img-thumbnail" height="300" width="410"
src="summarizerGraph?id=${$('#measure', page).val()}${request_parameter}&amp;aggregation=${$('#aggregation', page).val()}${randomParam}">`
                });
            }
            $('.del_img', page).click(function () {
                grid[pageIndex].removeWidget(this.parentNode.parentNode);
            });
        });

        $('#tabList').find('a').eq(pageIndex).tab('show'); // very messy :(
        if ($('.grid-stack ul', page).length !== 0) {
            grid[pageIndex].enableMove(false);
            grid[pageIndex].enableResize(false);

            projectAction.getDashboardConfiguration($(page).attr('id'), function (data) {
                const json = JSON.parse(data.responseObject());
                $.each(json, function (index) {
                    if (json[index].dashboard === $(page).attr('id')) {
                        if (json[index].id === 'unittest_overview') {
                            grid[pageIndex].addWidget({
                                w: 3,
                                h: 2,
                                content: `<span class="del_img float-left" style="display: none">
<svg class="svg" style="fill: #ff0000" aria-hidden="true">
    <use href="${resURL}/plugin/font-awesome-api/sprites/solid.svg#times"/>
</svg></span>
<span class="chk_show float-right" style="display: none">
<input type="checkbox" title="show in project overview" checked="checked"/></span>
<a href="./testRunGraph?width=800&amp;height=585&amp;id=unittest_overview${randomParam}" data-lightbox="${$(page).attr('id')}">
<img class="img-thumbnail" height="300" width="410" src="./testRunGraph?width=410&amp;height=300&amp;id=unittest_overview${randomParam}"></a>`
                            });
                        } else {
                            grid[pageIndex].addWidget({
                                w: 3,
                                h: 2,
                                content: `<span class="del_img float-left" style="display: none">
<svg class="svg" style="fill: #ff0000" aria-hidden="true">
    <use href="${resURL}/plugin/font-awesome-api/sprites/solid.svg#times"/>
</svg></span>
<span class="chk_show float-right" style="display: none">
<input type="checkbox" title="show in project overview" ${json[index].show ? "checked='checked'" : ''}/></span>
<a href="./summarizerGraph?width=800&amp;height=585&amp;id=${json[index].id}${randomParam}" data-lightbox="${$(page).attr('id')}" data-title="${json[index].description}">
<img class="img-thumbnail" height="300" width="410" 
src="./summarizerGraph?width=410&amp;height=300&amp;id=${json[index].id}${randomParam}" title="source: ${json[index].chartDashlet}-${json[index].measure} (${json[index].aggregation})
${json[index].description}"></a>`
                            });
                        }
                    }
                });
                $('.chk_show', page).hide();
                $('.del_img', page).hide().click(function () {
                    grid[pageIndex].removeWidget(this.parentNode.parentNode);
                });
            });
        }
        $('#donebutton', this).click(function () {
            const items = [];
            $('.grid-stack-item.ui-draggable', page).each(function () {

                const $this = $(this);
                items.push({
                    col: $this.attr('gs-x'),
                    row: $this.attr('gs-y'),
                    id: getURLParameter($('img', $this), 'id'),
                    dashboard: $(page).attr('id'),
                    chartDashlet: getURLParameter($('img', $this), 'chartDashlet'),
                    measure: getURLParameter($('img', $this), 'measure'),
                    customName: getURLParameter($('img', $this), 'customName'),
                    customBuildCount: getURLParameter($('img', $this), 'customBuildCount'),
                    show: $("input[type='checkbox']", $this).prop('checked'),
                    aggregation: getURLParameter($('img', $this), 'aggregation')
                });
            });
            projectAction.setDashboardConfiguration($(page).attr('id'), JSON.stringify(items), function () {
                location.reload(true);
            });
        });
    });

    const hash = window.location.hash;
    if (hash) {
        $(`ul.nav a[href="${hash}"]`).tab('show');
    } else {
        $('#tabList').find('a:first').tab('show'); // Select first tab
    }

    $('.nav-tabs a').click(function () {
        $(this).tab('show');
        const scrollMem = $('body').scrollTop() || $('html').scrollTop();
        window.location.hash = this.hash;
        $('html,body').scrollTop(scrollMem);
    });
})(jQuery3);

function generateTitle(measure, chartDashlet, aggregation) {
    const chartDashletName = measure.replace(/\s/g, '') === chartDashlet.replace(/\s/g, '') ? chartDashlet : `${chartDashlet} - ${measure}`;
    return `${chartDashletName} (${aggregation})`;
}

function getURLParameter(obj, parameter) {
    return $(obj).attr('src').indexOf(parameter) > -1 ? wurl(`?${parameter}`, $(obj).attr('src')) : ''
}

function encode(toEncode) {
    return encodeURIComponent(toEncode)
        .replace(/!/g, '%21')
        .replace(/'/g, '%27')
        .replace(/\(/g, '%28')
        .replace(/\)/g, '%29')
        .replace(/\*/g, '%2A');
}
